### PR TITLE
Adds feature specification for workers

### PR DESCRIPTION
**Pull request by nkremerh**

### CODE REVIEW

**Comment by nkremerh** [comment]
My main concern here is that there is some implementation-specific code we placed in here. @btovar, could you please take a look at this to help me determine which changes need to be made to make this generic?

File: work_queue/src/perl/Work_Queue/Task.pm [hunk 1]
```diff
@@ -58,6 +58,11 @@ sub specify_category {
 	return work_queue_task_specify_category($self->{_task}, $name);;
 }
 
+sub specify_requirement {
```
Comment (btovar) [inline work_queue/src/perl/Work_Queue/Task.pm]: property  or feature  instead of requirement?

File: work_queue/src/work_queue.c [hunk 2]
```diff
@@ -2611,6 +2618,28 @@ static work_queue_msg_code_t process_resource( struct work_queue *q, struct work
...
+
+	if(n != 1)
+		return MSG_FAILURE;
+
+	if(!w->features)
+		w->features = hash_table_create(4,0);
+
+	url_decode(feature, fdec, WORK_QUEUE_LINE_MAX);
+
+	debug(D_WQ, "FEATURE FOUND: %s\n", fdec);
```
Comment (btovar) [inline work_queue/src/work_queue.c]: No caps.

File: work_queue/src/work_queue.c [hunk 3]
```diff
@@ -2611,6 +2618,28 @@ static work_queue_msg_code_t process_resource( struct work_queue *q, struct work
...
 }
 
+static work_queue_msg_code_t process_feature( struct work_queue *q, struct work_queue_worker *w, const char *line )
+{
+	char feature[WORK_QUEUE_LINE_MAX];
+	char fdec[WORK_QUEUE_LINE_MAX];
+
+	int n = sscanf(line, "feature %s", feature);
+
+	if(n != 1)
```
Comment (btovar) [inline work_queue/src/work_queue.c]: Use { } even for single line ifs.

File: work_queue/src/work_queue.c [hunk 4]
```diff
@@ -3991,6 +4049,15 @@ struct work_queue_task *work_queue_task_clone(const struct work_queue_task *task
 	new->command_line = xxstrdup(task->command_line);
   }
 
+  if(task->user_resources) {
```
Comment (btovar) [inline work_queue/src/work_queue.c]: Using features, requirements and user_resources for the same thing. Choose one.

File: work_queue/src/work_queue.c (hunk 3 above)
Comment (nkremerh) [reply work_queue/src/work_queue.c]: I agree.

**Comment by nkremerh** [comment]
@btovar Streamlined the naming conventions and fixed a couple formatting issues along with a bug. Ready for a second pass?

File: doc/man/work_queue_worker.m4 [hunk 5]
```diff
@@ -70,6 +70,7 @@ OPTION_PAIR(--gpus, n)Set the number of GPUs this worker should use. (default=0)
 OPTION_PAIR(--memory, mb)Manually set the amount of memory (in MB) reported by this worker.
 OPTION_PAIR(--disk, mb)Manually set the amount of disk space (in MB) reported by this worker.
 OPTION_PAIR(--wall-time, s)Set the maximum number of seconds the worker may be active.
+OPTION_PAIR(--feature, feature)Specifies a user-defined feature the worker provides.
```
Comment (btovar) [inline doc/man/work_queue_worker.m4]: add in parenthesis '(can be repeated)'

Comment (nkremerh) [reply doc/man/work_queue_worker.m4]: Done.

File: work_queue/src/work_queue_worker.c [hunk 6]
```diff
@@ -1931,6 +1931,33 @@ static int serve_master_by_hostport( const char *host, int port, const char *ver
 
 	measure_worker_resources();
 
+	//Check GPU name if we can
+	char *gpu_name;	
+	int fd = open("/bin/nvidia-smi", O_RDONLY);
```
Comment (btovar) [inline work_queue/src/work_queue_worker.c]: Nate,

Please see   dttools/src/cctools_gpu_autodetect

Is that tool doing the same things? If so, please consolidate (remove the script, or remove this c code).

**Pull request merged**